Fix Game starts with 1 player on ready

rob-v opened the pull request:
1. rename Session.IsSinglePlayer to IsSingleHuman for clarity - this is optional, 11 of 12 cases seem to be correct, 1 case where it isn't is fixed in commit 2
2. Lobby - updated auto start on 'Ready' checkbox and on 'Start Game' button to fix case when on dedicated non SinglePlayer server game could be started with only 1 human player and admin as spectator.

~This PR solves this specific case: admin as player or spectator is ready. One of 2 players moves to spectator slot. At this moment 'all' = 1 player is ready and game starts. As it is immediate, this player might not notice and starts playing without opponent.
In general this PR prevents starting game by changing 'Ready' checkbox if there aren't at least 2 players (humans or bots). In case admin might really want to start a game with only 1 player or bot, he can still click 'Start Game'.~

rob-v commented:
As discussed, this is not good idea, so closing.

pchote commented:
There is a real bug here, which is that servers will autostart if the there is only one player, even if `server.LobbyInfo.GlobalSettings.EnableSingleplayer` is enabled, if the admin is spectating and ready.

rob-v commented:
Don't we want to allow 1 admin spectating 1 human player vs 1 bot on dedicated server?

pchote commented:
No.  We don't allow one human player spectating one admin vs 1 bot on a dedicated server, and this case is no different.

rob-v commented:
Updated - please see PR's description.
I think MapGlobal.IsSinglePlayer should be also updated:
		[Desc("Returns true if there is only one human player.")]
		public bool IsSinglePlayer { get { return Context.World.LobbyInfo.IsSingleHuman; } }
even without renaming IsSinglePlayer, there is something wrong - either it should really check for only 1 human player or description isn't precise (True also when only 1 human as spectator)

chrisforbes reviewed (approved):
Not tested, but this looks OK.

pchote commented on OpenRA.Game/Network/Session.cs:
```diff
@@ -90,11 +90,16 @@ public string FirstEmptyBotSlot()
 			get { return Clients.Count(c => c.Bot == null) == 1; }
 		}
 
+		public IEnumerable<Session.Client> HumanPlayers
```
Using "Human" here doesn't sit well with me.  There is no requirement for the other client to be human, just that it is another real game instance and not a bot ran on the same client.

How about `NonBotPlayers`?

rob-v replied:
There is already another HumanPlayers property https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Game/GameInformation.cs#L38. (non bot players). Do you wish to rename both or only this new?

pchote replied:
One thing at a time.  That affects external compatibility, so is a much bigger deal to change.

pchote commented on OpenRA.Game/Network/Session.cs:
```diff
@@ -90,11 +90,16 @@ public string FirstEmptyBotSlot()
 		}
 
-		public bool IsSinglePlayer
+		public bool IsSingleHuman
```
IMO this property should go away.  A lot of code uses this as a proxy for "one human player", which is wrong, because this will return false if there is one human player plus another human observer.

I suggest introducing an extra `NonBotClients` property that works like `HumanPlayers`/`NonBotPlayers` but doesn't filter spectators.  We can then replace all the uses of `IsSinglePlayer` with the appropriate version, using `.Count()` inline instead of introducing more mostly redundant methods.

pchote commented on OpenRA.Game/Network/UnitOrders.cs:
```diff
@@ -118,7 +118,7 @@ public static void ProcessOrder(OrderManager orderManager, World world, int clie
 						{
 							var pause = order.TargetString == "Pause";
-							if (orderManager.World.Paused != pause && world != null && !world.LobbyInfo.IsSinglePlayer)
+							if (orderManager.World.Paused != pause && world != null && !world.LobbyInfo.IsSingleHuman)
```
I don't see why we shouldn't show this for singleplayer games.

pchote commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -375,7 +375,7 @@ void ValidateClient(Connection newConn, string data)
 					client.Name, newConn.Socket.RemoteEndPoint);
 
-				if (Dedicated || !LobbyInfo.IsSinglePlayer)
+				if (Dedicated || !LobbyInfo.IsSingleHuman)
```
This could then simplify down to `if (LobbyInfo.NonBotClients.Count() > 1)` (IMO there's no reason to show the player their own join message on dedicated servers)

pchote commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -392,7 +392,7 @@ void ValidateClient(Connection newConn, string data)
 				}
 
-				if (!LobbyInfo.IsSinglePlayer && Map.DefinesUnsafeCustomRules)
+				if (!LobbyInfo.IsSingleHuman && Map.DefinesUnsafeCustomRules)
```
Checking the player count here is a bug: it suppresses the message if somebody is joining an empty server that defaults to a map with custom rules.

pchote commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -679,7 +679,7 @@ public void StartGame()
 
 			// HACK: Turn down the latency if there is only one real player
-			if (LobbyInfo.IsSinglePlayer)
+			if (LobbyInfo.IsSingleHuman)
```
This can use `LobbyInfo.NonBotClients.Count() > 1`.

rob-v replied:
You mean LobbyInfo.NonBotClients.Count() == 1?

pchote replied:
I do, yes.

pchote commented on OpenRA.Game/Traits/Player/DeveloperMode.cs:
```diff
@@ -113,7 +113,7 @@ public DeveloperMode(DeveloperModeInfo info)
 		void INotifyCreated.Created(Actor self)
 		{
-			Enabled = self.World.LobbyInfo.IsSinglePlayer || self.World.LobbyInfo.GlobalSettings
+			Enabled = self.World.LobbyInfo.IsSingleHuman || self.World.LobbyInfo.GlobalSettings
```
Spectators shouldn't be counted here, so `LobbyInfo.NonBotPlayers.Count() > 1`

rob-v replied:
You mean LobbyInfo.NonBotPlayers.Count() == 1?

pchote replied:
I do indeed.

pchote commented on OpenRA.Mods.Common/Scripting/Global/MapGlobal.cs:
```diff
@@ -108,7 +108,7 @@ public string TerrainType(CPos cell)
 
 		[Desc("Returns true if there is only one human player.")]
-		public bool IsSinglePlayer { get { return Context.World.LobbyInfo.IsSinglePlayer; } }
+		public bool IsSinglePlayer { get { return Context.World.LobbyInfo.IsSingleHuman; } }
```
Similarly here: Having lua scripts behave differently for the same single *player* game depending on whether or not there are spectators is a bug.  `LobbyInfo.NonBotPlayers.Count() == 1`

pchote commented on OpenRA.Mods.Common/ServerTraits/PlayerPinger.cs:
```diff
@@ -36,7 +36,7 @@ public void Tick(S server)
 
 				// Ignore client timeout in singleplayer games to make debugging easier
-				if (server.LobbyInfo.IsSinglePlayer && !server.Dedicated)
+				if (server.LobbyInfo.IsSingleHuman && !server.Dedicated)
```
This does care about spectators, so `NonBotClients.Count() < 2`

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs:
```diff
@@ -47,7 +47,7 @@ public IngameChatLogic(Widget widget, OrderManager orderManager, World world, Mo
 
 			var players = world.Players.Where(p => p != world.LocalPlayer && !p.NonCombatant && !p.IsBot);
-			disableTeamChat = world.IsReplay || world.LobbyInfo.IsSinglePlayer || (world.LocalPlayer != null && !players.Any(p => p.IsAlliedWith(world.LocalPlayer)));
+			disableTeamChat = world.IsReplay || world.LobbyInfo.IsSingleHuman || (world.LocalPlayer != null && !players.Any(p => p.IsAlliedWith(world.LocalPlayer)));
```
Again `NonBotPlayers`, considering spectators here is a bug.

pchote replied:
Actually I think we can drop the check here completely.  The filtered `players` check already makes sure that you have two humans on the same team.

rob-v replied:
It disabled Team chat in case of 1 player or 1 observer. For 1 observer it won't be disabled by removing IsSinglePlayer.

pchote reviewed (commented):
This has opened up a small can of worms:

Commit:
Fix IsSinglePlayer

rob-v commented:
Updated

abcdefg30 commented on OpenRA.Mods.Common/ServerTraits/PlayerPinger.cs:
```diff
@@ -36,7 +36,7 @@ public void Tick(S server)
 
 				// Ignore client timeout in singleplayer games to make debugging easier
-				if (server.LobbyInfo.IsSinglePlayer && !server.Dedicated)
+				if (server.LobbyInfo.NonBotClients.Count() < 2 && !server.Dedicated)
```
Is it even possible to have less then one non-bot client? (Otherwise `NonBotClient.Count() == 1` would work as well, wouldn't it?) Anyway, just wondering, this shouldn't make any difference in practice.

rob-v replied:
I kept it as suggested by pchote. It isn't possible in OpenRA IMO atm, but in theory there could be bot only game simulated.

abcdefg30 commented on OpenRA.Game/Network/UnitOrders.cs:
```diff
@@ -118,7 +118,7 @@ public static void ProcessOrder(OrderManager orderManager, World world, int clie
 						{
 							var pause = order.TargetString == "Pause";
-							if (orderManager.World.Paused != pause && world != null && !world.LobbyInfo.IsSinglePlayer)
+							if (orderManager.World.Paused != pause && world != null)
```
Can we get the check here back? I find it quite annoying to get two messages each time you open and close the menu.

rob-v replied:
Done

abcdefg30 commented on OpenRA.Mods.Common/Scripting/Global/MapGlobal.cs:
```diff
@@ -108,7 +108,7 @@ public string TerrainType(CPos cell)
 
 		[Desc("Returns true if there is only one human player.")]
-		public bool IsSinglePlayer { get { return Context.World.LobbyInfo.IsSinglePlayer; } }
+		public bool IsSinglePlayer { get { return Context.World.LobbyInfo.NonBotPlayers.Count() == 1; } }
```
Although this is in line with the description: Do we really want to count bot-only matches as non-singleplayer? Imho we should rather adjust the description and use `NonBotPlayers.Count() == 1`.

rob-v replied:
You mean NonBotClients.Count() == 1?

abcdefg30 replied:
Er, yeah.

rob-v replied:
Done, reverted to previous condition using the new property.

abcdefg30 reviewed (commented):
Looks good otherwise.

rob-v commented:
Updated.

abcdefg30 commented:
@chrisforbes does you 👍 still stand?

obrakmann commented on OpenRA.Mods.Common/Scripting/Global/MapGlobal.cs:
```diff
@@ -107,8 +107,8 @@ public string TerrainType(CPos cell)
-		[Desc("Returns true if there is only one human player.")]
-		public bool IsSinglePlayer { get { return Context.World.LobbyInfo.IsSinglePlayer; } }
+		[Desc("Returns true if there is only one human client.")]
+		public bool IsSinglePlayer { get { return Context.World.LobbyInfo.NonBotClients.Count() == 1; } }
```
I think we actually want to have `NonBotPlayers` here. The presence of potential specs won't be of any interest to scripts.

rob-v replied:
Ok, desc reverted and used NonBotPlayers.

abcdefg30 replied:
It actually is: https://github.com/OpenRA/OpenRA/pull/13271#discussion_r117537334. I think it is a bug to count bot-only matches as "multiplayer" matches. So either `NonBotClients` or `NonBotPlayers < 2`.

obrakmann reviewed (dismissed):
Looks fine, just one minor nit.

rob-v commented:
Updated.
